legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/domain/api/test/TestModelBuilder.java

```diff
@@ -92,17 +92,30 @@ public TestModelBuilder(DependenciesApi dependenciesApi, EntityApi entityApi, Pr
     {
         try
         {
+            Set<String> latestUpstreamDependenciesIds = SetAdapter.adapt(latestUpstreamDependencies).collect(dependency -> dependency.getDepotProjectId().toString());
+            if (latestUpstreamDependenciesIds.contains(downstreamProjectId))
+            {
+                throw new IllegalArgumentException("Project " + downstreamProjectId + " was specified as downstream but in fact it is a direct dependency for upstream project " + upstreamProjectId.toString());
+            }
+
             //get transitive dependencies for upstream project from metadata
+            Set<DepotProjectVersion> latestUpstreamTransitiveDependencies = Sets.mutable.empty();
             latestUpstreamDependencies.forEach(depotProjectVersion ->
             {
                 Set<DepotProjectVersion> projects = this.metadataApi.getProjectDependencies(depotProjectVersion.getDepotProjectId().toString(), depotProjectVersion.getVersionId(), true);
                 // remove upstream project as it has changed
                 projects.removeIf(project -> project.getDepotProjectId().equals(upstreamProjectId));
-                latestUpstreamDependencies.addAll(projects);
+                latestUpstreamTransitiveDependencies.addAll(projects);
             });
```

Comment:
You can use flatCollect to build this set more elegantly. Also, I think it's going to be better to remove the upstream project once at the end, rather than each time.